### PR TITLE
Implement `seal_code_hash` and `seal_own_code_hash`

### LINKED ISSUE #1204: Add `code_hash` and `code_own_hash`
This is to remember to implement these functions as a follow-up to [substrate#10933](https://github.com/paritytech/substrate/pull/10933/)

### PR DESCRIPTION
**Pull request by agryaznov**
Follow-up for [substrate#10933](https://github.com/paritytech/substrate/pull/10933).
Resolves #1204.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 ⚠️ **The ink! `master` is ahead of your branch, this might skew the comparison data below.**

These are the results when building the `examples/*` contracts from this branch with `cargo-contract 1.1.0-ef194a0` and comparing them to ink! `master`: 

|                          | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:--------------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`            |                  |            | 1.03 K               |                |
| `adder`                  |                  |            | 2.17 K               |                |
| `contract-introspection` |                  |            | 2.37 K               |                |
| `contract-terminate`     |                  |            | 0.94 K               | 214_418        |
| `contract-transfer`      |                  |            | 8.15 K               | 14_418         |
| `delegate-calls`         |                  |            | 2.99 K               | 14_950         |
| `delegator`              |                  |            | 6.37 K               | 46_295         |
| `dns`                    |                  |            | 8.84 K               | 43_254         |
| `erc1155`                |                  |            | 17.23 K              | 86_508         |
| `erc20`                  |                  |            | 8.49 K               | 43_254         |
| `erc721`                 |                  |            | 11.81 K              | 115_344        |
| `flipper`                |                  |            | 1.31 K               | 14_418         |
| `forward-calls`          |                  |            | 2.90 K               | 29_536         |
| `incrementer`            |                  |            | 1.21 K               | 14_418         |
| `multisig`               |                  |            | 25.20 K              | 93_527         |
| `rand-extension`         |                  |            | 3.92 K               | 14_418         |
| `subber`                 |                  |            | 2.19 K               |                |
| `trait-erc20`            |                  |            | 8.76 K               | 43_254         |
| `trait-flipper`          |                  |            | 1.00 K               | 14_418         |
| `trait-incrementer`      |                  |            | 1.19 K               | 28_836         |
| `upgradeable-flipper`    |                  |            | 1.55 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/186437) | Last update: Fri Apr  1 14:20:30 CEST 2022

File: crates/env/src/api.rs [hunk 1]
```diff
@@ -514,6 +514,22 @@ where
     })
 }
 
+/// Retrieves the code hash of a contract living at specified account
```
Comment (cmichi) [inline crates/env/src/api.rs:517]:
```suggestion
/// Retrieves the code hash of the contract at the specified account id.
```

Comment (cmichi) [inline crates/env/src/api.rs:517]: Could you also add a section `/// # Errors` here?

File: crates/env/src/api.rs [hunk 2]
```diff
@@ -514,6 +514,22 @@ where
...
+pub fn code_hash<E>(account: &E::AccountId) -> Result<E::Hash>
+where
+    E: Environment,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| instance.code_hash::<E>(account))
+}
+
+/// Retrieves the code hash of the currently executing contract.
```
Comment (cmichi) [inline crates/env/src/api.rs:525]: Could you also add a section `/// # Errors` here?

File: crates/env/src/backend.rs [hunk 3]
```diff
@@ -447,4 +447,23 @@ pub trait TypedEnvBackend: EnvBackend {
     fn caller_is_origin<E>(&mut self) -> bool
     where
         E: Environment;
+
+    /// Retrieves the code hash of the contract for the given `account`,
```
Comment (cmichi) [inline crates/env/src/backend.rs:451]:
```suggestion
    /// Retrieves the code hash of the contract at the given `account` id,
```

File: crates/env/src/engine/off_chain/impls.rs [hunk 4]
```diff
@@ -486,4 +486,18 @@ impl TypedEnvBackend for EnvInstance {
     {
         unimplemented!("off-chain environment does not support cross-contract calls")
     }
+
+    fn code_hash<E>(&mut self, _account: &E::AccountId) -> Result<E::Hash>
+    where
+        E: Environment,
+    {
+        unimplemented!("off-chain environment does not support contract instantiation")
```
Comment (cmichi) [inline crates/env/src/engine/off_chain/impls.rs:494]:
```suggestion
        unimplemented!("off-chain environment does not support `code_hash`")
```

File: crates/env/src/engine/off_chain/impls.rs [hunk 5]
```diff
@@ -486,4 +486,18 @@ impl TypedEnvBackend for EnvInstance {
...
+    }
+
+    fn own_code_hash<E>(&mut self) -> Result<E::Hash>
+    where
+        E: Environment,
+    {
+        unimplemented!("off-chain environment does not support contract instantiation")
```
Comment (cmichi) [inline crates/env/src/engine/off_chain/impls.rs:501]:
```suggestion
        unimplemented!("off-chain environment does not support `own_code_hash`")
```

File: crates/lang/src/env_access.rs [hunk 6]
```diff
@@ -901,4 +901,68 @@ where
     pub fn caller_is_origin(self) -> bool {
         ink_env::caller_is_origin::<E>()
     }
+
+    /// Returns the code hash of the contract living at the given `account`.
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:905]:
```suggestion
    /// Returns the code hash of the contract at the given `account` id.
```

**Review by cmichi — CHANGES_REQUESTED** [review]

File: crates/env/src/api.rs [hunk 7]
```diff
@@ -514,6 +514,30 @@ where
     })
 }
 
+/// Retrieves the code hash of the contract at the specified account id.
+///
+/// # Errors
+///
+/// If no code hash found for the specified account id.
```
Comment (cmichi) [inline crates/env/src/api.rs:521]:
```suggestion
/// Returns `Error::KeyNotFound` if no code hash was found for the specified account id.
```

File: crates/env/src/backend.rs [hunk 8]
```diff
@@ -447,4 +447,23 @@ pub trait TypedEnvBackend: EnvBackend {
     fn caller_is_origin<E>(&mut self) -> bool
     where
         E: Environment;
+
+    /// Retrieves the code hash of the contract at the given `account` id,
+    /// and stores the result in `output`.
```
Comment (cmichi) [inline crates/env/src/backend.rs:452]:
```suggestion
    /// Retrieves the code hash of the contract at the given `account` id.
```

File: crates/lang/src/env_access.rs [hunk 9]
```diff
@@ -901,4 +901,68 @@ where
     pub fn caller_is_origin(self) -> bool {
         ink_env::caller_is_origin::<E>()
     }
+
+    /// Returns the code hash of the contract at the given `account` id.
+    ///
+    /// # Example
+    ///
+    /// ```
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:909]:
```suggestion
    /// ```rust
```

File: crates/lang/src/env_access.rs [hunk 10]
```diff
@@ -901,4 +901,68 @@ where
...
+    ///
+    /// # Note
+    ///
+    /// For more details visit: [`ink_env::code_hash`]
+    pub fn code_hash(self, account_id: &E::AccountId) -> Result<E::Hash> {
+        ink_env::code_hash::<E>(account_id)
+    }
+
+    /// Returns the code hash of the contract living at the given `account`.
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:937]:
```suggestion
    /// Returns the code hash of the contract at the given `account` id.
```

File: crates/lang/src/env_access.rs [hunk 11]
```diff
@@ -901,4 +901,68 @@ where
...
+    ///
+    /// # Example
+    ///
+    /// ```
```
Comment (cmichi) [inline crates/lang/src/env_access.rs:941]:
```suggestion
    /// ```rust
```

File: crates/lang/src/env_access.rs (hunk 9 above)
Comment (agryaznov) [reply crates/lang/src/env_access.rs:909]: hm, why? and what about other examples?

File: crates/lang/src/env_access.rs (hunk 11 above)
Comment (agryaznov) [reply crates/lang/src/env_access.rs:941]: why?

File: crates/lang/src/env_access.rs (hunk 9 above)
Comment (cmichi) [reply crates/lang/src/env_access.rs:909]: Just checked and if no annotation is used, the language is Rust by default for markdown and `cargo test --doc`. So you can forget about this comment.

We already use `/// ```rust` in a number of doc comments, but let's just leave it as is there.

File: crates/env/src/engine/on_chain/impls.rs [hunk 12]
```diff
@@ -512,4 +512,28 @@ impl TypedEnvBackend for EnvInstance {
...
+    }
+
+    fn own_code_hash<E>(&mut self) -> Result<E::Hash>
+    where
+        E: Environment,
+    {
+        let output = &mut self.scoped_buffer().take(32);
+        ext::own_code_hash(output);
+        let hash = scale::Decode::decode(&mut &output[..])?;
+        Ok(hash)
```
Comment (cmichi) [inline crates/env/src/engine/on_chain/impls.rs:537]:
```suggestion
        scale::Decode::decode(&mut &output[..])
```

File: crates/env/src/engine/on_chain/impls.rs [hunk 13]
```diff
@@ -512,4 +512,28 @@ impl TypedEnvBackend for EnvInstance {
...
+        E: Environment,
+    {
+        let mut scope = self.scoped_buffer();
+        let output = scope.take(32);
+        scope.append_encoded(account_id);
+        let enc_account_id = scope.take_appended();
+
+        ext::code_hash(enc_account_id, output.as_mut())?;
+        let hash = scale::Decode::decode(&mut &output[..])?;
+        Ok(hash)
```
Comment (cmichi) [inline crates/env/src/engine/on_chain/impls.rs:527]:
```suggestion
        scale::Decode::decode(&mut &output[..])?
```

Comment (agryaznov) [reply crates/env/src/engine/on_chain/impls.rs:527]: Nah, this won't compile

```
expected enum `error::Error`, found struct `parity_scale_codec::Error`
```

**Review by cmichi — APPROVED** [review]

**Pull request merged**